### PR TITLE
Inject possible dependencies in codec and validator

### PR DESCRIPTION
**Pull request by bburnichon**
Make it possible to configure options on validators used

### CODE REVIEW

**Comment by webmozart** [comment]
Hi @bburnichon, thank you for this PR! :) What's the use case for this change?

**Comment by bburnichon** [comment]
I have a silex project where I need to configure specific validation for each encoding/decoding. This way, library is fully decoupled and I can have a configured JsonDecoder/JsonEncoder in Pimple container for each use case.

Also, I already have a `JsonSchema\UriRetriever` in my container and want to use it for all validations to minimize Http roundtrips. I inject my `UriRetriever` in `$validatorFactory`:

```php
$app['json-schema.retriever'] = $app->share(function () {
  return new UriRetriever();
});
$app['json-schema.validator_factory'] = $app->protect(function () use ($app) {
  return new Validator($app['json-schema.url_retriever']);
});

$app['json.validator'] = $app->share(function (Application $app) {
  return new JsonValidator($app['json-schema.validator_factory']);
});

$app['json.decoder'] = $app->share(function (Application $app) {
  return new JsonDecoder($app['json.validator']);
});
```

File: src/JsonValidator.php [hunk 1]
```diff
@@ -35,6 +35,33 @@ class JsonValidator
...
+     * Callable creating new Validator instances.
+     *
+     * @var callable
+     */
+    private $validatorFactory;
+
+    /**
+     * JsonValidator constructor.
+     *
+     * @param callable|null $validatorFactory Factory creating JsonSchema\Validator instances.
```
Comment (webmozart) [inline src/JsonValidator.php]: Why do you need support for callables in here instead of directly injecting the Validator?

Comment (bburnichon) [inline src/JsonValidator.php]: Because a new instance of Validator should be used at each call to check.

Comment (webmozart) [inline src/JsonValidator.php]: Why don't you simply `reset()` the validator before using it?

Comment (webmozart) [inline src/JsonValidator.php]: (which, of course, I could have done too :) )

Comment (bburnichon) [inline src/JsonValidator.php]: Because I did not knew it exists! ;)

**Comment by bburnichon** [comment]
This PR includes all changes from #9.

@webmozart I can squash commits if you prefer to keep a clean history.

File: src/JsonValidator.php [hunk 2]
```diff
@@ -135,4 +152,14 @@ private function loadSchema($file)
 
         return $schema;
     }
+
+    /**
+     * @return Validator
+     */
+    private function prepareValidator()
```
Comment (webmozart) [inline src/JsonValidator.php]: Thanks for updating! :) I don't think this needs to live in a separate method. You can simply inline the `reset()` call above.

**Comment by webmozart** [comment]
No, looks good :) Thanks a lot @bburnichon!

**Pull request merged**